### ISSUE TITLE
Snapshot detected as corrupted because of incorrect checksum comparison

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/camunda#19984]
**Describe the bug**
On restart of broker
We see 
```
Expected snapshot /usr/local/zeebe/data/raft-partition/partitions/3/snapshots/90926-1-821751-758053 to have checksum 1612185047, but the actual checksum is 3972871514; the snapshot is most likely corrupted. The startup will fail if there is no other valid snapshot and the log has been compacted.
```
then the Partition cannot startup
```
java.lang.IllegalStateException: In partition 2 expected to find a snapshot at index >= log's first index 211099, but found snapshot 0. A previous snapshot is most likely corrupted.
	at io.atomix.raft.utils.StateUtil.verifySnapshotLogConsistent(StateUtil.java:33) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:220) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:249) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:223) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:179) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.partition.impl.RaftPartitionServer.<init>(RaftPartitionServer.java:98) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.partition.RaftPartition.createServer(RaftPartition.java:133) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.partition.RaftPartition.initServer(RaftPartition.java:116) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.atomix.raft.partition.RaftPartition.bootstrap(RaftPartition.java:96) ~[zeebe-atomix-cluster-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup(RaftBootstrapStep.java:34) ~[zeebe-broker-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup(RaftBootstrapStep.java:14) ~[zeebe-broker-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.proceedWithStartupSynchronized(StartupProcess.java:162) ~[zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3(StartupProcess.java:170) ~[zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) ~[zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:96) ~[zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.5.4.jar:8.5.4]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.5.4.jar:8.5.4]
```

On inspecting further, the snapshot is not corrupted. But it is a bug in checksum comparison. 

**Environment:**
- Zeebe Version: 8.5.4

**Comment by Zelldon** [comment]
Related slack thread https://camunda.slack.com/archives/CSQ2E3BT4/p1719992603616489
Related incident channel https://app.slack.com/client/T0PM0P1SA/C07AMRPAVNE

**Comment by EuroLew** [comment]
The issue was when file chunks are received in the `FileBasedReceivedSnapshot.applyInternal` it will update the `checksumCollection` with `updateFromBytes` from every file this will result in a `.checksum` file getting persisted which has a combined checksum which is reliant on reading the contents of the file as `updateFromBytes` updates the checksum with the contents of the file.

When the broker restarts and then checks the snapshot for validity, it calculates the checksums of the files but this time using a `ChecksumProvider` avoiding reading the entire file therefore the combined checksum will be different.

The individual file checksums will still be the same that is why in all the other patches a combined checksum equality check was substituted for a checksum equality verification (every file checksum is equal). But this was missed in the 8.5 backport.

The PR which fixes this removes the the combined checksum verification.

**Comment by lenaschoenburg** [comment]
@EuroLew Could you check why this wasn't caught in tests?

**Comment by EuroLew** [comment]
> @EuroLew Could you check why this wasn't caught in tests?

There just wasn't a test that 
- Received a snapshot (snapshots taken locally use the `ChecksumProvider` so combined and individual file checksums will match)
- Restarted the broker (upon which it will verify the files match some expected checksum, in this case we change the check to verifying each file is as expected instead of some combined value).

**Comment by Zelldon** [comment]
But this should happen on our benchmarks right? It was easily reproducible in SaaS

**Comment by lenaschoenburg** [comment]
> But this should happen on our benchmarks right? It was easily reproducible in SaaS

Quite possibly the 8.5 release benchmark never had to do snapshot replication.

**Comment by deepthidevaki** [comment]
After the investigation the root cause is more clear. This happens in only the following situation:

* A follower receives a snapshot from the leader
* Follower restart after that, before it takes a local snapshot

Probably that's why we did not observe. Even if there was a snapshot replication, if the broker do no restart after that there is no problem.

**Comment by deepthidevaki** [comment]
Follow up for @EuroLew after the fix is released.

- Check if we have to fix the checksum calculation in `FilebasedReceivedSnapshot`
- Can we write unit tests in snapshot module to cover this case?

**Comment by EuroLew** [comment]
> Follow up for @EuroLew after the fix is released.
> 
> * Check if we have to fix the checksum calculation in `FilebasedReceivedSnapshot`

It is possible to fix so that combined check matches up however we do not use it for anything right now and the aim is to deprecate it so it should probably not be done.

> * Can we write unit tests in snapshot module to cover this case?

It is possible.
- create files
- receivedSnapshot.applyInternal all the files
- persist the corresponding snapshot
- `FileBasedSnapshotStore.restore` 

this would trigger the same check.

**Comment by deepthidevaki** [comment]
@EuroLew Let's do the followup asap. We can sync tomorrow.

**Comment by lenaschoenburg** [comment]
@deepthidevaki & @EuroLew 
This error popped up again during a backup restore: https://console.cloud.google.com/errors/detail/CP279Oesof_bdw;service=zeebe;time=P7D;locations=global?project=camunda-saas-dev

I think this is another case where we should have had tests. Basically:
1. Take a backup.
2. Restore from it.

Could you either add such a test or check why an existing test didn't catch this bug?

**Comment by deepthidevaki** [comment]
Had a quick look at it. It is because the restore is not using the correct the checksum builder. 

- We should remove the constructor that accepts null ChecksumProvider to avoid accidental misuse. 
- I think unit tests wouldn't cover this. So extend qa restore tests where the backup contains snapshot.

**Comment by deepthidevaki** [comment]
Let's open a new issue and fix it asap. @EuroLew Do you have capacity for this?

**Comment by lenaschoenburg** [comment]
I think this isn't a new issue and _not_ providing a ChecksumProvider should be fine, during restore that just means that we calculate all full file checksums and don't rely on ones provided by RocksDB. I think the new error is just because the comparison of checksums was broken in 8.5.4.

**Comment by EuroLew** [comment]
> Let's open a new issue and fix it asap. @EuroLew Do you have capacity for this?

sure will do this

**Comment by deepthidevaki** [comment]
I think I do not fully understand the issue then. Based on Lena's comment, it seems it is already fixed. But it is just a matter for providing new test cases to cover this case. Is that right?

**Comment by lenaschoenburg** [comment]
Yes, to me it seems like restore was completely broken in 8.5.4 due to the same bug we already fixed here. I was wondering how we didn't notice that. If it's a missing test, we should definitely add one!

**Comment by EuroLew** [comment]
Maybe I'm confused about our release process and how customers get our patch fixes. 

we merged the fix into 8.5/stable but the logs show that the fix hasn't been merged it is still comparing combined checksums. But then the error says its occuring on 8.5.4  what exactly is going on here? Does merging the patch fix into 8.5/stable mean 8.5.4 customers don't get the fix?

**Comment by deepthidevaki** [comment]
? Does merging the patch fix into 8.5/stable mean 8.5.4 customers don't get the fix?

The fix is release in patch 8.5.5. If the cluster is still on 8.5.4, they are still affected by this bug. They have to update to 8.5.5.

**Comment by EuroLew** [comment]
> ? Does merging the patch fix into 8.5/stable mean 8.5.4 customers don't get the fix?
> 
> The fix is release in patch 8.5.5. If the cluster is still on 8.5.4, they are still affected by this bug. They have to update to 8.5.5.

ah ok that clears up some confusion I had.

**Comment by EuroLew** [comment]
> Yes, to me it seems like restore was completely broken in 8.5.4 due to the same bug we already fixed here. I was wondering how we didn't notice that. If it's a missing test, we should definitely add one!

In 8.5.4 all internal snapshots are through a `FileBasedSnapshotStore` that uses a RocksDB Checksum Provider but the `PartitionRestoreService` when doing a `.restore` uses a store without a checksum provider so now it now there will be a mismatch. Weird a test didn't catch this, it would just need to 
1. take a backup like how prod would
2. call the partition restore service .restore

unit tests aren't expected to catch this as it would require making rocksDB snapshots work in the test but integration tests should've caught this I'll write one.

**Comment by nicpuppa** [comment]
@lenaschoenburg @deepthidevaki a similar issue happened again on a 8.6.5 cluster. The error message is slightly different:

```
Uncaught exception in 'Startup' in phase 'STARTED'. Continuing with next job.

java.lang.IllegalStateException: In partition 1 expected to find a snapshot at index >= log's first index 172456941, but found snapshot 0. A previous snapshot is most likely corrupted.
	at io.atomix.raft.utils.StateUtil.verifySnapshotLogConsistent(StateUtil.java:33) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:226) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:249) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:223) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:180) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.partition.impl.RaftPartitionServer.<init>(RaftPartitionServer.java:99) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.partition.RaftPartition.createServer(RaftPartition.java:133) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.partition.RaftPartition.initServer(RaftPartition.java:116) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.atomix.raft.partition.RaftPartition.bootstrap(RaftPartition.java:96) ~[zeebe-atomix-cluster-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup(RaftBootstrapStep.java:34) ~[zeebe-broker-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup(RaftBootstrapStep.java:14) ~[zeebe-broker-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.proceedWithStartupSynchronized(StartupProcess.java:162) ~[zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3(StartupProcess.java:170) ~[zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) ~[zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:85) ~[zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:42) [zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.6.5.jar:8.6.5]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.6.5.jar:8.6.5]
```

Should I re-open this issue, or it's better to open a new one ? The cluster was able to restart properly tho. Here the [logs](https://cloudlogging.app.goo.gl/MnoNF4FM43HKY2t5A) and the [data folder](https://drive.google.com/drive/folders/1oF0wMS8N29TFES20Thdj9hGK8hOBE-kn)

**Comment by npepinpe** [comment]
While the partition ID is different, it sounds like pretty much the same thing to me 🤔

**Comment by npepinpe** [comment]
Are we having yet another critical issue, in an older version? What's up with this week, third corruption issue (but all in old versions)

**Comment by korthout** [comment]
I just encountered [the same error group](https://console.cloud.google.com/errors/detail/CI6K3a-Tl-vrygE;service=zeebe;time=P7D;locations=global?project=camunda-saas-prod) during my medic rotation as @nicpuppa reported above. I've linked it to this issue now.

@npepinpe I'll reopen this issue, because of your comment [here](https://github.com/camunda/camunda/issues/19984#issuecomment-3562337312) rather than opening a new one:
>While the partition ID is different, it sounds like pretty much the same thing to me 🤔

Note that the related cluster did not encounter any problems due to the error:
- The affected broker was a FOLLOWER for all three partitions
- It simply restarted
- Other brokers were unaffected
- Quorum was maintained for all partitions

<img width="1127" height="309" alt="Image" src="https://github.com/user-attachments/assets/6a122250-9d5b-4ad4-b4a8-850ae3800f24" />

While this issue may be labeled `severity/critical`, the impact on this encounter was not one of data loss.

**Comment by deepthidevaki** [comment]
I had a quick check with the shared data folder - The broker (version 8.6.5) was able to start Partition 1 correctly. Snapshot is not corrupted.

**Comment by korthout** [comment]
@deepthidevaki Does that mean we can lower the severity?

**Comment by deepthidevaki** [comment]
@korthout Based on your last comment - if it recovered automatically after a restart, it is not a critical bug. We can lower the severity. @nicpuppa Was it similar in your case too?

**Comment by korthout** [comment]
>@nicpuppa Was it similar in your case too?

It's the same case, the error group was simply not acknowledged/linked

**Comment by korthout** [comment]
Lowering to high severity, as the impact was noticable (restarted broker)

**Comment by nicpuppa** [comment]
yes @deepthidevaki , I didn't acknowledged the error waiting for feedback 👍

**Comment by korthout** [comment]
@nicpuppa For critical issues I think it's better to consider the worst and just take action. We can always lower severity once we know more.

**Comment by deepthidevaki** [comment]
Then it sounds a like a different issue. I would suggest to open a new one.

**Comment by nicpuppa** [comment]
> For critical issues I think it's better to consider the worst and just take action. We can always lower severity once we know more.

It didn't look critical because the cluster was able to recover, that's why I thought that waiting on feedback on how to properly proceed was better.

**Comment by npepinpe** [comment]
Yes, let's open a new one, since the impact is different.

**Comment by korthout** [comment]
I've opened a new issue:
- https://github.com/camunda/camunda/issues/41354

Closing this one again.
